jupyter_server/services/kernels/v3/client.py

```diff
@@ -0,0 +1,733 @@
+    shell_channel_class = Type(ShellChannel)
+    control_channel_class = Type(ControlChannel)
+    stdin_channel_class = Type(StdinChannel)
+
+    # Set of listener functions - don't use Traitlets Set, just plain Python set
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._listeners = {}  # Maps callback -> filter config
+        self._listening = False
+
+        # Connection state tracking
+        self._connecting = False
+        self._connection_ready = False
+        self._connection_ready_event = asyncio.Event()
+
+        # Message queue for messages received before connection is ready
+        self._queued_messages = []
+        self._max_queue_size = 1000  # Prevent memory issues
+
+        # Note: The session is already EncodedMsgIdSession, created by the KernelManager
+        # No need to replace it here
+
+    def add_listener(
+        self,
+        callback: t.Callable[[str, list[bytes]], None],
+        msg_types: t.Optional[t.List[t.Tuple[str, str]]] = None,
+        exclude_msg_types: t.Optional[t.List[t.Tuple[str, str]]] = None,
+    ):
+        """Add a listener to be called when messages are received.
+
+        Args:
+            callback: Function that takes (channel_name, msg_bytes) as arguments
+            msg_types: Optional list of (msg_type, channel) tuples to include.
+                      If provided, only messages matching these filters will be sent to the listener.
+                      Example: [("status", "iopub"), ("execute_reply", "shell")]
+            exclude_msg_types: Optional list of (msg_type, channel) tuples to exclude.
+                              If provided, messages matching these filters will NOT be sent to the listener.
+                              Example: [("status", "iopub")]
+
+        Note:
+            - If both msg_types and exclude_msg_types are provided, msg_types takes precedence
+            - If neither is provided, all messages are sent (default behavior)
+        """
+        if msg_types is not None and exclude_msg_types is not None:
+            raise ValueError("Cannot specify both msg_types and exclude_msg_types")
+
+        # Store the listener with its filter configuration
+        self._listeners[callback] = {
+            "msg_types": set(msg_types) if msg_types else None,
+            "exclude_msg_types": set(exclude_msg_types) if exclude_msg_types else None,
+        }
+
+    def remove_listener(self, callback: t.Callable[[str, list[bytes]], None]):
+        """Remove a listener."""
+        self._listeners.pop(callback, None)
+
+    def mark_connection_ready(self):
+        """Mark the connection as ready and process queued messages."""
+        if not self._connection_ready:
+            self._connecting = False
+            self._connection_ready = True
+            self._connection_ready_event.set()
+
+            # Process queued messages
+            asyncio.create_task(self._process_queued_messages())
```

Comment:
Should we track these as background tasks for proper cleanup if the client shuts down?

```python
task = asyncio.create_task(self._process_queued_messages())
if not hasattr(self, '_background_tasks'):
    self._background_tasks = []
self._background_tasks.append(task)
```